packages/transaction-messages/src/fee-payer.ts

```diff
@@ -3,23 +3,39 @@ import { Address } from '@solana/addresses';
 import { BaseTransactionMessage } from './transaction-message';
 
 export interface ITransactionMessageWithFeePayer<TAddress extends string = string> {
-    readonly feePayer: Address<TAddress>;
+    readonly feePayer: Readonly<{ address: Address<TAddress> }>;
 }
 
 export function setTransactionMessageFeePayer<
     TFeePayerAddress extends string,
-    TTransaction extends BaseTransactionMessage,
+    TTransactionMessage extends BaseTransactionMessage & Partial<ITransactionMessageWithFeePayer>,
 >(
     feePayer: Address<TFeePayerAddress>,
-    transaction: TTransaction,
-): ITransactionMessageWithFeePayer<TFeePayerAddress> & Omit<TTransaction, 'feePayer'> {
-    if ('feePayer' in transaction && feePayer === transaction.feePayer) {
-        return transaction as ITransactionMessageWithFeePayer<TFeePayerAddress> & Omit<TTransaction, 'feePayer'>;
+    transactionMessage: TTransactionMessage,
+): ITransactionMessageWithFeePayer<TFeePayerAddress> & Omit<TTransactionMessage, 'feePayer'> {
+    if (
+        'feePayer' in transactionMessage &&
+        isAddressOnlyFeePayer(transactionMessage.feePayer) &&
+        feePayer === transactionMessage.feePayer.address
```

Comment:
Tiny nit: I usually do the cheap checks first then the more elaborate ones, because of how these expressions short circuit on a `false` outcome.